samples/bluetooth/handsfree_ag/src/main.c

```diff
@@ -0,0 +1,403 @@
+/* main.c - Application main entry point */
+
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/classic/rfcomm.h>
+#include <zephyr/bluetooth/classic/sdp.h>
+#include <zephyr/bluetooth/classic/hfp_ag.h>
+#include <zephyr/settings/settings.h>
+
+static void connected(struct bt_conn *conn, uint8_t err);
+static void disconnected(struct bt_conn *conn, uint8_t reason);
+static void security_changed(struct bt_conn *conn, bt_security_t level, enum bt_security_err err);
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed,
+};
+
+static void ag_connected(struct bt_hfp_ag *ag);
+static void ag_disconnected(struct bt_hfp_ag *ag);
+static void ag_sco_connected(struct bt_hfp_ag *ag, struct bt_conn *sco_conn);
+static void ag_sco_disconnected(struct bt_hfp_ag *ag);
+static void ag_ringing(struct bt_hfp_ag *ag, bool in_band);
+static void ag_accept(struct bt_hfp_ag *ag);
+static void ag_reject(struct bt_hfp_ag *ag);
+static void ag_terminate(struct bt_hfp_ag *ag);
+static void ag_outgoing(struct bt_hfp_ag *ag, const char *number);
+static void ag_incoming(struct bt_hfp_ag *ag, const char *number);
+
+static struct bt_hfp_ag_cb ag_cb = {
+	.connected = ag_connected,
+	.disconnected = ag_disconnected,
+	.sco_connected = ag_sco_connected,
+	.sco_disconnected = ag_sco_disconnected,
+	.outgoing = ag_outgoing,
+	.incoming = ag_incoming,
+	.ringing = ag_ringing,
+	.accept = ag_accept,
+	.reject = ag_reject,
+	.terminate = ag_terminate,
+};
+
+static struct bt_conn *default_conn;
+struct bt_hfp_ag *hfp_ag;
+
+static struct bt_br_discovery_param br_discover;
+
+static struct bt_br_discovery_param br_discover;
+static struct bt_br_discovery_result scan_result[CONFIG_BT_HFP_AG_DISCOVER_RESULT_COUNT];
+
+struct k_work discover_work;
+struct k_work_delayable call_connect_work;
+struct k_work_delayable call_disconnect_work;
+
+struct k_work_delayable call_remote_ringing_work;
+struct k_work_delayable call_remote_accept_work;
+
+NET_BUF_POOL_DEFINE(sdp_discover_pool, 10, BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU),
+		    CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+
+static uint8_t sdp_discover_cb(struct bt_conn *conn, struct bt_sdp_client_result *result);
+
+static struct bt_sdp_discover_params sdp_discover = {
+	.func = sdp_discover_cb,
+	.pool = &sdp_discover_pool,
+	.uuid = BT_UUID_DECLARE_16(BT_SDP_HANDSFREE_SVCLASS),
+};
```

Comment:
Might be possible here too. Generally, if you can avoid forward declarations by rearranging the static functions, then that's always preferred.

Reply:
updated